pkg/autoscaler/metrics/http_scrape_client_test.go

```diff
@@ -250,3 +279,61 @@ func newTestHTTPClient(response *http.Response, err error) *http.Client {
 		},
 	}
 }
+
+type benchmarkingRoundTripper struct {
+	dataType benchmarkingDataType
+}
+
+func (brt benchmarkingRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	return makeProtoResponseForBenchmarking(http.StatusOK, brt.dataType, network.ProtoAcceptContent), nil
+}
+
+func newBenchmarkingtHTTPClient(dataType benchmarkingDataType) *http.Client {
+	return &http.Client{
+		Transport: benchmarkingRoundTripper{dataType: dataType},
+	}
+}
+
+type benchmarkingDataType int
+
+const (
+	NoData benchmarkingDataType = iota
+	SmallData
+	MediumData
+	LargeData
+)
+
+func BenchmarkUnmarshallingProtoData(b *testing.B) {
+	benchmarks := []struct {
+		name     string
+		dataType benchmarkingDataType
+	}{{
+		name:     "BenchmarkStatWithEmptyPodName",
+		dataType: NoData,
+	}, {
+		name:     "BenchmarkStatWithSmallPodName",
+		dataType: SmallData,
+	}, {
+		name:     "BenchmarkStatWithMediumPodName",
+		dataType: MediumData,
+	}, {
+		name:     "BenchmarkStatWithLargePodName",
+		dataType: LargeData,
+	}}
+
+	for _, bm := range benchmarks {
+		hClient := newBenchmarkingtHTTPClient(bm.dataType)
+		scrapeClient, err := newHTTPScrapeClient(hClient)
+		if err != nil {
+			b.Fatal(err)
+		}
+		b.Run(bm.name, func(b *testing.B) {
+			b.ReportAllocs()
+			for i := 0; i < b.N; i++ {
+				if _, err := scrapeClient.Scrape(testURL); err != nil {
```

Comment:
I'm wondering if this would all be a little easier if we only benchmarked the `statFromProto` function to not have everything skewed by the http client etc.

Reply:
Ok I can do that.